lib/active_merchant/billing/gateways/payu_latam.rb

```diff
@@ -0,0 +1,484 @@
+
+        additional_values = {}
+        additional_values[:TX_VALUE] = tx_value
+        additional_values[:TX_TAX] = tx_tax
+        additional_values[:TX_TAX_RETURN_BASE] = tx_tax_return_base
+
+        post[:transaction][:order][:additionalValues] = additional_values
+      end
+
+      def add_signature(post)
+        post[:transaction][:order][:signature] = signature_from(post)
+      end
+
+      def signature_from(post)
+        signature_string = [
+          @options[:api_key],
+          @options[:merchant_id],
+          post[:transaction][:order][:referenceCode],
+          post[:transaction][:order][:additionalValues][:TX_VALUE][:value],
+          post[:transaction][:order][:additionalValues][:TX_VALUE][:currency]
+        ].compact.join("~")
+
+        Digest::MD5.hexdigest(signature_string)
+      end
+
+      def add_payment_method(post, payment_method, options)
+        if payment_method.is_a?(String)
+          _, brand, token = split_authorization(payment_method)
+          credit_card = {}
+          credit_card[:securityCode] = options[:cvv] if options[:cvv]
+          post[:transaction][:creditCard] = credit_card
+          post[:transaction][:creditCardTokenId] = token
+          post[:transaction][:paymentMethod] = brand.upcase
+        else
+          credit_card = {}
+          credit_card[:number] = payment_method.number
+          credit_card[:securityCode] = payment_method.verification_value
+          credit_card[:expirationDate] = format(payment_method.year, :four_digits).to_s + '/' + format(payment_method.month, :two_digits).to_s
+          credit_card[:name] = payment_method.name.strip
+          post[:transaction][:creditCard] = credit_card
+          post[:transaction][:paymentMethod] = BRAND_MAP[payment_method.brand.to_s]
+        end
+      end
+
+      def add_payer(post, options)
+        payer = {}
+        payer[:dniNumber] = options[:payer_dni_number] if options[:payer_dni_number]
+        payer[:cnpj] = options[:payer_cnpj] if options[:payer_cnpj]
+        payer[:emailAddress] = options[:email] || 'unspecified@example.com'
+        if address = options[:billing_address]
+          post[:transaction][:paymentCountry] = address[:country]
+          payer[:fullName] = address[:name]
+          payer[:contactPhone] = address[:phone]
+          billing_address = {}
+          billing_address[:street1] = address[:address1]
+          billing_address[:street2] = address[:address2]
+          billing_address[:city] = address[:city]
+          billing_address[:state] = address[:state]
+          billing_address[:country] = address[:country]
+          billing_address[:postalCode] = address[:zip]
+          billing_address[:phone] = address[:phone]
+          payer[:billingAddress] = billing_address
+        end
+        post[:transaction][:payer] = payer
+      end
+
+      def add_browser_data(post, options)
+        post[:transaction][:deviceSessionId] = 'n/a'
```

Comment:
Do the transactions fail if `deviceSessionId`, `cookie`, and `userAgent` aren't specified.  If so, the 👍 on filling them the with `n/a`.